src/burst_buffer/dws.lua

```diff
@@ -0,0 +1,227 @@
+function DWS:apply(fname)
+	return self:io_popen("kubectl apply -f " .. fname .. " 2>&1")
+end
+
+-- DWS.delete will delete the named Workflow resource via kubectl.
+-- On success this returns true.
+-- On failure this returns false and the output of the kubectl command.
+function DWS:delete()
+	return self:io_popen("kubectl delete workflow " .. self.name .. " 2>&1")
+end
+
+-- DWS.get_jsonpath will get the specified values from the Workflow resource.
+-- On success this returns true and a string containing the result of the query.
+-- On failure this returns false and the output of the kubectl command.
+function DWS:get_jsonpath(path)
+	local cmd = "kubectl get workflow " .. self.name .. " -o jsonpath='" .. path .. "' 2>&1"
+	return self:io_popen(cmd)
+end
+
+-- DWS.get_current_state will get the status of the Workflow resource with
+-- respect to its desired state.
+-- On success this returns true and an array containing the desiredState,
+-- the current state, and the status of the current state.
+-- On failure this returns false and the output of the kubectl command.
+function DWS:get_current_state()
+	local ret, output = self:get_jsonpath([[{.spec.desiredState}{"\n"}{.status.state}{"\n"}{.status.status}{"\n"}]])
+	if ret == false then
+		return ret, output
+	end
+	local status = {}
+	local idx = 0
+	for line in output:gmatch("[^\n]+") do
+		status[idx] = line
+		idx = idx + 1
+	end
+	if idx == 1 and status[0] == "proposal" then
+		-- DWS has not yet attached a status section to this new
+		-- resource.  Fill out the table for our consumer.
+		status[1] = ""
+		status[2] = ""
+	elseif idx ~= 3 then
+		-- We should have had 3 lines.
+		ret = false
+	end
+	return ret, status
+end
+
+-- DWS.get_hurry will get the hurry flag of the Workflow resource.
+-- On success this returns true and a boolean for the value of the hurry flag.
+-- On failure this returns false and the output of the kubectl command.
+function DWS:get_hurry()
+	local ret, result = self:get_jsonpath("{.spec.hurry}")
+	if ret ~= true then
+		return ret, result
+	end
+	local hurry = false
+	if result == "true" then
+		hurry = true
+	end
+	return true, hurry
+end
+
+-- DWS.wait_for_status_complete will loop until the workflow reports that
+-- its status is completed.
+-- On success this returns true and the status text.
+-- On failure this returns false and an error message.
+function DWS:wait_for_status_complete(max_passes)
+	while max_passes > 0 do
+		local done, output = self:get_current_state()
+		if done == false then
+			return false, output
+		end
+		-- Wait for current state to reflect the desired state, and
+		-- for the current state to be complete.
+		if output[0] == output[1] and string.find(output[2], "Completed") ~= nil then
+			return true, output
+		elseif string.find(output[2], "Error") ~= nil then
+			return false, output
+		end
+		os.execute("sleep 1")
+		max_passes = max_passes - 1
+	end
+	return false, "exceeded max wait time"
```

Comment:
See: https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#wait

You can use the following command to skip the while/sleep cycle. I believe you can look for two values (e.g. "Completed,Error"), but the online documentation is a bit murky. 
```shell
kubectl wait --timeout=30 --for=jsonpath='{.status.status}'= Completed,Error ...
```

After getting one of those two statuses, you can do a single call to "get_current_state".

Reply:
I spent some time going through the possibilities, trying to see how to incorporate this.  The caller would have to pass in the desiredState to this function, so that the --for=jsonpath could be built to look at .status.state and expect it to equal the pass-in desiredState.   So that's one wait for --timeout=max_passes.  Then I'd have to do a second wait for max_passes until .status.status is Completed.  Then I'd have to do a third query to see all of .spec.desiredState, .status.state, and .status.status to verify that they line up as expected.

So now max_passes is grossly fuzzy.

I just don't think that kubectl wait is a win here.

Reply:
Can you kick off both `kubectl wait` commands in the background and grab their pids. Then do a `wait -n pid1 pid2`. Then you can make another call to get the status.

Reply:
Even if we were using a language that didn't force me to jump through hoops to do that, I'm not sure how that would be a better solution.  Can you explain more about why you feel that it's important to build the solution on kubectl wait?

Reply:
You're right. It's not important at this point.